Baystation12 Rigsuits

tigercat2000 opened the pull request:
Ports Bay's fancy rigsuit stuff.
Doesn't replace our own hardsuits, it's just an addition.

tigercat2000 commented:
I'll note off the bat, in this comment, to not detract from the OP (or something)

Almost all of this system is completely inaccessible to players.
There is no way to obtain new modules.
No way to obtain new rigsuits.

TheDZD commented:
Just to clear things up, what is the difference CE/engi/regular rigsuits and the current ones?

Markolie commented:
That little room needs:
1) Its own area if it doesn't have it.
2) An APC.
3) Scrubber/vent/air alarm.
4) Light switch.
5) Firelock if it doesn't have it.
6) Should probably use the blue "C" door.
7) Fire alarm.

It also looks a bit bare, consider putting down some floor decals.

tigercat2000 commented:
@TheDZD
Most importantly, these things:
 - Rigsuits are contained in boxes, which equip to your back. You can then expand these boxes into an exosuit, boots, gloves, and a helmet.
 - Rigsuits have special modules
  - Specifically, for each class; CE has a plasma cutter, intellicard system, built-in jetpack, mesons
  - Engineer has a welding fuel tank which can refuel welding tools in hand, built-in jetpack, and mesons
  - Plain EVA suit has a jetpack and mesons
 - Some rigsuits require access on your ID in order to interact with their UI or extend/seal/retract them.
 - Extending above, all rigsuits have a nanoUI system, which lists all of their modules, as well as core buttons like sealing the suit and such
 - Rigsuits require a charge in their contained cell, if that cell dies, the rigsuit will go into depowered mode, which can range from slowdown to complete blindness and slowdown. If the suit dies, you are basically unable to get out of it until you can power it again.
  - I am actually missing bay's "hardsuit removal" surgery, which I only just realized, so that will be a way to get out of a dead suit when I can be bothered.
 - There is probably more I am forgetting

TheDZD commented:
Main things I have issue with would then be the fact that you can be permanently stuck in a rigsuit unless it gets surgery'ed off. I'm assuming the CE plasma cutter acts like a fuel-less welding tool, not like a ranged weapon like they do currently? Either way, I'm kind of conflicted on that, as it is power creep, especially considering how many cool toys the already CE starts out with.

tigercat2000 commented:
@TheDZD Actually, uh.. It is technically the plasma gun, but do to the way rig modules work, it can't fire, only attack walls, but not people. I just realized it will run out of ammo with the way I have it setup, another thing for the growing todo list.

The rigsuit being stuck isn't that big of a deal, really; The only thing that takes power normally is sealing/unsealing and the AI moving your suit, until you get into the special modules. 
Even then, the choice is pretty much between "get stuck in the suit" or "suit stops functioning because bay's systems are broken even on bay, and you can rip the suit off but it still has NODROP therefore it kinda glitches the fuck out and eventually breaks to the point where you have to delete every piece indivudally"

tigercat2000 commented:
Updated room: ![biwa](https://puu.sh/kxEUR/e74ce1c02b.png)

Robertha89 commented:
actually, if you could leave the walls to space there as regular walls instead of reinforced ones, that would be great. It's one of the few areas you can sneakily break into the ai upload from.

tigercat2000 commented:
@Robertha89 that isn't consistant with the rest of the station and not something I intend to do

KasparoVy commented:
Would it be possible to tailor the rigsuit system such that it detects the user's species and deploys a suit-worn sprite accordingly? Like, a Tajaran backpacks the rig, deploys, BAM. Tajaran rigsuit appropriate to the job closest tied to the rig (engineer or CE), ditto for Unathi, Vox, Skrell

Or are the deployed rig sprites different-- like would more have to be made, or are the worn rig sprites the same and it's just the addition of the SYSTEM (read as: functionality and the rig deployment system)?

tigercat2000 commented:
@KasparoVy that's already a thing, the rigsuits use species sprite sheets

KasparoVy commented:
Wicked! That's a relief. Thanks for the response.

Fox-McCloud commented:
Absolutely not. :-1:  x :100: 

Know up front that my face is set against this and I will likely disagree with any/all implementations of this that are anywhere remotely accessible to the crew.

A rigsuit that can't be removed, is armored out the ass, is super quick compared to regular hardsuits, not to mention has undisarmable mounted weapons is INSANE power creep. It's literally power armor 2.0 with a fancier package and more fleshed out features.  That alone is bad---on top of this you deliberately added in the AI and pAI control despite the fact two maintainers specifically told you "no" on it.

This also returns fully armored people walking around with flash protection, THICKMATERIAL, identity hiding, built in mesons, etc with no downsides other than "needs to recharge a cell on occasion".

This also adds in a duplicate NASA void suit despite the fact we already have one---and sure enough, it has insanely ridiculous armor stats; I see no point to all the void suits, as they're just super buffed spaces suits, which we deliberately nerfed and made difficult to access because of their ridiculously powergamey nature. That's NOT something we need back.

Not only this, but the amount of snowflake checks just to get this to work is worrying, such as tint checks, human movement, etc.

Lastly, you're altering the recharge mechanics of things--which isn't good. We're solidly in line with TG's gun system (and continuing go head in that direction)--while I realize some things listed are there for coding efficiency's sake and less lines of code, it's still altering the behavior of recharge--particularly for the borg weapons.

Just no no no. Sorry, no :-1:

Spacemanspark commented:
:+1: I want this.

tigercat2000 commented:
@Fox-McCloud 
Okay, I would love to just accept your input, but I'm going to point out the flaws here.
> A rigsuit that can't be removed, is armored out the ass, is super quick compared to regular hardsuits, not to mention has undisarmable mounted weapons is INSANE power creep.

The admin rigsuits, yeah, they are armored and super quick, because they are admin rigsuits.
However, the 4 crew accessible rigsuits aren't that way.
The engineering hardsuit, for example, has a slowdown equivalent to a diona. It also has the *exact* armor stats as our current hardsuit.
The CE hardsuit, yeah, I can't argue too much, it is 100% an advantage over the CE normal suit. However, it has identical armor stats.
RD hardsuit has the same armor stats as the CE rigsuit and hardsuit, with a minor buff to radiation.
I believe the medical hardsuits share the engineering hardsuit values.

The can't remove part, I can understand is a bit of a sore point, but these are so ridiculously rare I don't think it's a problem.
Not a single player-accessible hardsuit has undisarmable mounted weapons; The closest they come is the plasma gun, which I couldn't get in testing to actually harm a mob.

>This also returns fully armored people walking around with flash protection, THICKMATERIAL, identity hiding, built in mesons, etc with no downsides other than "needs to recharge a cell on occasion".

How is this any different than normal hardsuits? Mesons are incredibly easy to come by, and they have a huge downside by the fact that they can't use gloves or a backpack while wearing the suit. 

>This also adds in a duplicate NASA void suit despite the fact we already have one---and sure enough, it has insanely ridiculous armor stats; I see no point to all the void suits, as they're just super buffed spaces suits, which we deliberately nerfed and made difficult to access because of their ridiculously powergamey nature. That's NOT something we need back.

The voidsuits are entirely inaccessible, and are only there to make sure everything functions properly. I could probably rip them out with no problem, but it could cause unexpected issues.

>Lastly, you're altering the recharge mechanics of things--which isn't good. We're solidly in line with TG's gun system (and continuing go head in that direction)--while I realize some things listed are there for coding efficiency's sake and less lines of code, it's still altering the behavior of recharge--particularly for the borg weapons.

All the recharge thing is, is a unified proc for guns, with snowflaking for hardsuits in it. Nothing is functionally different, it just reduces the copypasted code and shoves in a snowflake check.

> That alone is bad---on top of this you deliberately added in the AI and pAI control despite the fact two maintainers specifically told you "no" on it.

No, I ported the system 1:1 as best as I could, then fixed everything that was broken. I'm not fully opposed to removing it, but everything people have said they have a problem with on it seems to be fueled by misinformation. The most common things I have heard, "but this will let them control dead bodies!", "remote controlled hardsuits via AI is shit", and "Entirely tanks with no human restrictions" are absolutely untrue. 
You can't control a hardsuit if the body inside it is stunned, weakened, resting, dead, lying, buckled, paralyzed, in crit, etc etc etc. Even when you can, you drain cells incredibly quickly, and the user can instantly turn it off. Not to mention, you can't fire weapons, or do literally anything but basic cardinal movement and toggle resting. Remote-control isn't a thing either, the only thing that can control your hardsuit is the IIS module, which you have to grab an MMI, pAI, or AI and put in your suit. 

If your opinion stays the same, so be it, but I think you are working based on a bias of knowing how it works on bay then my implementation.

TheDZD commented:
Not really a fan of the RD having his own personal rigsuit, nor there being any more medical EVA gear than there already is. That said, you should probably make the medical rigsuits match the armor values of the medical hardsuits.

Zuhayr commented:
Rigs are currently bugged on Baycode due to issues with canremove, wouldn't recommend porting them while they're broken as shit.

tigercat2000 commented:
@Zuhayr I noticed, and fixed most of the issues anyways. We can't use canremove, we got rid of it in favor of the NODROP flag, so I adapted it and fixed the system up.

Zuhayr commented:
Ah cool. No worries then. Did you grab the newer/better interface code from /vg/ as well? If not, it might be worth looking at.

tigercat2000 commented:
@Zuhayr I didn't, but the basic nanoUI is pretty damn nice.

Zuhayr commented:
I believe it's included in https://github.com/d3athrow/vgstation13/tree/hardsuit_port

Fox-McCloud commented:
>The CE hardsuit, yeah, I can't argue too much, it is 100% an advantage over the CE normal suit. However, it has identical armor stats.
RD hardsuit has the same armor stats as the CE rigsuit and hardsuit, with a minor buff to radiation.
I believe the medical hardsuits share the engineering hardsuit values.

Things worth pointing out:

The Industrial hardsuit has a welding tank with 300 fuel in it that automatically refuels welders when it's activated and one is in your hand--this seems problematic to me, particularly for Blobs.

CE hardsuit has no slowdown, the same armor, a free undisarmable weapon, not to mention literally no one can wear it but the CE and Captain.

RD hardsuit has lower slowdown, increased armor over engineering, both in terms of radiation and energy weapons), and only the RD and Captain can wear it....and yet it still has flash protection. Side note: the anomally scanner is literally worthless/functionless. 

The Medical hardsuit has flash protection when it absolutely shouldn't; the whole point of the medical hardsuit was that it traded off increased maneuverability for less combat protection...furthemore, it has an injector that can inject a ton of healing chems, on demand (by the by, with the exception of nutriment, spaceacillin, and radium, none of those chems exist anymore).

 
The EVA hardsuits are probably the worst though; there's practically no downsides to not wearing one of these. It has no slowdown, THICKMATERIAL, flash protection, a built in light, meson vision, and jetpack. Compare this to light EVA suits which are barely armored and have no flash protection in exchange for being slightly faster than hardsuits--they confer no advantage other than pressure and cold temperature protection.

>The can't remove part, I can understand is a bit of a sore point, but these are so ridiculously rare I don't think it's a problem.

It means you can never discover the identity of someone who refuses to remove their suit--it means you can't space them (after knocking them out) as an effective means of killing them--it means memorizing/wiki'ing yet another shitty Bay surgery just to remove a hardsuit off of someone so you can treat them--I'm sorry, but a surgeon shouldn't have to put up with yet another surgery just to begin treating someone effectively.

>The voidsuits are entirely inaccessible, and are only there to make sure everything functions properly. I could probably rip them out with no problem, but it could cause unexpected issues.

There's no need for them; we already hard traditional hard and spacesuits for literally all those things--there's really no sense in adding a ton of suits for people to clamor for when we have the existing ones that are already largely balanced out and fit for our purposes.

>All the recharge thing is, is a unified proc for guns, with snowflaking for hardsuits in it. Nothing is functionally different, it just reduces the copypasted code and shoves in a snowflake check.

It's not the same, you increased the cost of borg weapons by TEN fold. Admittedly, this wouldn't be a problem if someone didn't screw with cell ratings a long time ago.

>You can't control a hardsuit if the body inside it is stunned, weakened, resting, dead, lying, buckled, paralyzed, in crit, etc etc etc. Even when you can, you drain cells incredibly quickly,

You seem to contract yourself. So, which is it, could you explain please? You seem to suggest you can't, but then turn around and suggest that you, in fact, can under some scenarios. I've never once imagined that it was remote control---I always understood it to be "load PAI/AI into hardsuit->can control your movements after you're dead/unconscious"....and the code seems to suggest just that: at least for AI's, at the very least; that movements can be controlled, even after the person is dead or even just paralyzed--the latter of which is even more ridiculous as it renders anyone in a suit like this with an AI as completely immune to every single knockout chem---their little digital pal just takes over and moves them away from danger. If that truly is the case, that's ridiculous overpowered on laughable levels.

Earthdivine commented:
I am still very iffy on the Eva and CE ones, but from staff chat it sounded like engineers won't abuse theirs due to all the penalties.

I would recommend giving the lethal presets to the SST's as they have been needing an upgrade for a long while and it would help make them scarier, but I don't know how good of an idea that is.

Markolie commented:
I certainly think this is a good system to have a place if we can get rid of the snowflake code (the vision bit in life.dm especially is terrible, there has to be a better way). Whether or not the crew should have any access to these suits at all I'm not sure of, but it's certainly something nice to give to the SST like Earth suggested and the Deathsquad.

Zuhayr commented:
@Fox-McCloud Just a note, the surgery is target chest, click with welder, it unseals the suit so it can be removed. It's not a multi-step procedure or hard to memorize.

@Markolie I believe Psi is working on a vision refactor for Bay that should help with the snowflakes in vision code. Might be worth hitting him up on IRC.

Fox-McCloud commented:
@Zuhayr 

I know, but it's yet another extra "thing" that makes it more difficult to remove these for seemingly no reason at all other than to make it more difficult to do so. It's making a quick security check of "let's see who this asshole is who's been giving people hell" into an ordeal that either medial has to get involved in or someone from security has to end up getting a welder and tabling them to resolve.....provided they don't rest-unrest spam to prevent you from doing so to begin with.

tigercat2000 commented:
>You seem to contract yourself. So, which is it, could you explain please? You seem to suggest you can't, but then turn around and suggest that you, in fact, can under some scenarios. I've never once imagined that it was remote control---I always understood it to be "load PAI/AI into hardsuit->can control your movements after you're dead/unconscious"....and the code seems to suggest just that: at least for AI's, at the very least; that movements can be controlled, even after the person is dead or even just paralyzed--the latter of which is even more ridiculous as it renders anyone in a suit like this with an AI as completely immune to every single knockout chem---their little digital pal just takes over and moves them away from danger. If that truly is the case, that's ridiculous overpowered on laughable levels. 

Alright, I can see how I might have sounded contradictory;
I meant, if the mob is disabled, and their client cannot move, you are not able to move either. When the mob is healthy and _can_ move, you can move it as well, if the AI control is there.
The forced_move proc is mostly a copypaste of client/Move() from what I understand; Since most movement is handled via client/Move(), there is no easy way to sanity-check everything for moving a mob by an external client.
The trick here is that the limitations are handled, frankly, all over the place; 
[This](https://github.com/ParadiseSS13/Paradise/pull/2278/files#diff-5597b35f1c3df45115d4e0dae367a094R829) handles the stat/weakened/paralyzed/chems; It checks that the mob wearing the suit passes canmove, which they won't if they are dead, weakened, paralyzed, stunned, lying, buckled, resting, and so on.
[This](https://github.com/ParadiseSS13/Paradise/pull/2278/files#diff-5597b35f1c3df45115d4e0dae367a094R781) handles AI's being unable to access suits remotely; They can only interact with the movement and UI while actually loaded into the suit, same goes for pAI's and MMI's.

Fox-McCloud commented:
Alright, after having tested this and getting help with @TheDZD from testing this, there's a few things:

- You can use modules even when cuffed; this includes point-blank shotting people with the plasma cutter or welding down walls/girders....or injecting chems into yourself with the medical suit
- the pAI/Ai insertion into a suit can be easily abused to make you run faster than you otherwise would.
- None of the helmets actually seem to have lights on them.
- The EVA suit is horrendously overpowered; no slowdown, free mesons, free magboots, armor, thickmaterial, free jetpack, and flash protection. 
- Free mesons is actually worse than you think; it not only gives you meson vision, but allows you to wear another pair of glasses with it, meaning you can do things like mesons+thermals/medical/sec (this applies to medical HUD on the medical rig as well).
- You can utilize any glove type you like with any suit--just retract the gauntlets; it's of no actual consequence.
- I already pretty much covered what I felt needs to be said on the RD hardsuit and the medical hardsuit---I really don't think these need to be in, anyway.
- Annoying thing: if you try to equip the rig to your back and move, it doesn't just cancel the equipping, it makes you drop it, entirely, for some reason.
- I still don't see the need for the back item to have NODROP when powered on and activated; if you manually remove this flag and yank it off, it gracefully unequips all items and powers the all down without any issues or runtimes.
- this seems to have broken the interface for intellicards and pAI's both; you can't do anything with pAI buttons on the menu, for instance; when an AI was loaded back into a card from a suit, it didn't even detect there was an AI in the card to begin with, despite everything else indicated as such.


Some of these things--particularly the balance issues, I take less issue with if these are never going to be accessible to players and remain admin/event/DS/SST only...if it's going to be the latter, then I'm fine with some of the more powergamey aspects remaining in place.

tigercat2000 commented:
Thanks, I think, A lot of that is useful.
I'll note on a few things, though;
> I still don't see the need for the back item to have NODROP when powered on and activated; if you manually remove this flag and yank it off, it gracefully unequips all items and powers the all down without any issues or runtimes.

It does *now* after I put some work into it. Bay's system didn't originally do that, it broke entirely. So, I suppose I could do that now, not sure I want to though.

>Annoying thing: if you try to equip the rig to your back and move, it doesn't just cancel the equipping, it makes you drop it, entirely, for some reason.

yeeeaaahh I noticed that.

>You can use modules even when cuffed; this includes point-blank shotting people with the plasma cutter or welding down walls/girders....or injecting chems into yourself with the medical suit

That sounds like a bug, but not sure. The interface should definitely be using typical nanostate.
About the mesons, if I recall correctly, the system just entirely breaks if you try to wear multiple glasses at once.

>None of the helmets actually seem to have lights on them

This is because of how bay handles head lights; It's absolutely fucking insane, I couldn't think of a way to implement it correctly.

Fox-McCloud commented:
>the system just entirely breaks if you try to wear multiple glasses at once.

wasn't my general experience; I was able to use built in mesons+a medical/sec HUD with ease.

In any event, as I said, if these are not accessible to players, then stuff like glasses stacking and such isn't really that important, and only the bugs are.

Zuhayr commented:
@tigercat2000 If you need an explanation of the head lights system I can help you out. It's not very transparent but it's just a cache system internally.

tigercat2000 commented:
Won't be able to finish this for a little while.

Aurorablade commented on code/modules/clothing/spacesuits/rig/suits/station.dm:
```diff
@@ -0,0 +1,229 @@
+/obj/item/clothing/head/helmet/space/new_rig/hazard
+
+/obj/item/weapon/rig/internalaffairs
+	name = "augmented tie"
```
yessssssssssssssssssssssssss......

PsiOmegaDelta commented:
https://github.com/Baystation12/Baystation12/pull/11219 does improve the rig-snowflake vision situation to a degree, but it's still not completely sorted.

Fox-McCloud commented:
@PsiOmegaDelta 

We already have a way to custom set tint and flash--we're mostly interested in things like SEE TURFS, SEE MOBS, etc.

PsiOmegaDelta commented:
That is also changed in the linked PR.

tigercat2000 commented:
I'll have to fix the intellicard compile issues later, after #2305 is merged.

Fox-McCloud commented:
Another thing worth pointing out.

Now that https://github.com/ParadiseSS13/Paradise/pull/2292 has been merged, having NODROP on these suits effectively means a person is literally impossible to search this inventory unless you give him the suit removal surgery (which he can resist, if he so desires).

Spacemanspark commented:
Perhaps make a Deathsquad RIG and give it to them exclusively?

Robertha89 commented:
Can't resist surgery after being cuffed.

Spacemanspark commented:
You can keep standing back up, the only way to keep you down is via force (stunning).

Fox-McCloud commented:
@Robertha89 

Yup, what @Spacemanspark said--if someone absolutely does not want to have surgery done to, they can keep rest-spamming just to prevent the surgery from even being started (even if cuffed or straight-jacked). The only way to prevent that is by keeping them permanently stun-locked.

Robertha89 commented:
True. But if they are being like that. Its easier to kill em, then remove it. 
I mean I like the concept of a collapsible spacesuit. Why not just make the surgery to remove it to remove power. Then crowbar it open...

I'd I would like the suits in cause interesting. But on the other hand.  The power creeping. But again that can be solved by a quick exucutions too. I am torn between having it or not... like sufficient access ID could make it collapse...

Fox-McCloud commented:
@Robertha89 

That would require even more snowflaking; there's no way to access an item on someone's back without removing it--likewise, attacking a person with an ID card only attacks the person and runs an armor check--it won't transfer any of that to the suit--for it to do so, you'd have to add in even MORE snowflake checks in human defense.

And the whole point of "it's easier to kill em, then remote it" presumes that they're deserving of death---players who've committed no-permanent crimes can and will give security trouble just to make their lives more difficult than they already are---you can't kill someone who's come in for trespassing just because you can't get their suit off.

Robertha89 commented:
Make it impossible to resist when cuffed then?

orelbon commented:
what's the difference between void suits and hard suits?

tigercat2000 commented:
@orelbon Voidsuits are basically our current hardsuits. They just happen to be unobtainable; Only kept them in the code at all because it might cause issues if they are missing.

Aurorablade commented:
heres a thought that might be rather silly, what about something like how the Mech have a console to lock them down or EMP them. What about a device that could be used to override the RIG and force it to unseal/retract? 'Remote Rig Override Device' I would imagine NT would install some kind of kill switch of such on the suits. (instead of being global like a console you have to be in a certain range of them on use)

Now this might be a disadvantage to traitors UNLESS they emagged it. Even then you don't want to disable to the RROD to the ground, so maybe an emagged suit would still be crippled by a RROD but still give the user the chance to run like hell and reboot/recover it. Silly silly silly idea.....

tigercat2000 commented:
@Aurorablade My current plans for this, when I get back to working on it (it's pissing me off too much to finish right now)
 
 - Make NODROP rig backs rely on a var, 'var/uses_NODROP' or something, which will be set to 0 for all player-accessible rigsuits
 - Fixes:
   - Intellicards being shit and broken as hell
   - RIG's staying in hand instead of dropping
   - Whatever the hell is causing the ability to use modules while stunned
   - Head lights
 - Maybe limit free mesons to CE rig only
 - Maybe rip out voidsuits, if I can find the time to thoroughly test that they are completely unnecessary
 - Port bay hardsuit surgery, Maybe, not necessarily necessary if I make the rigs non-NODROP.
 - Look into borg weapons being 10x cost (not intended at all)
 - Nerf the fuck out of EVA suits
 - Nerf the fuck out of medical suits
 - Consider fixing changing CE values, slowdown and armor. Not sure on this yet, it IS the CE armor afterall.
 - Probably nerf the fuck out of the RD suit if not remove it entirely

tigercat2000 commented:
Alright, going to update the OP in a moment, but I ripped out a few things:

 - Armor changes reverted.
 - Map changes reverted.
 - Merge conflicts fixed.

Basically, this is no longer player-accessible.
I keep finding projects that require elements of this, and ripping them out is only going to cause merge conflicts.
I'll do the rest of the fixes in a second.

tigercat2000 commented:
Shit I'm not going to fix until I find a time to make these player-usable:
 - NODROP shit
 - Hardsuit Surgery
 - Value tweaks
 - Extensive testing

tigercat2000 commented:
Going to ask for a full review of this from @Markolie / @Fox-McCloud / @TheDZD 

This should be merge-ready at this point, balance issues shouldn't be a problem given that they are admin-spawn only at this point. 

I do still intend to fix them up and play with numbers later, but these are making me very tired, and I have other projects that rely on core components of this- like some template work for the pool controller, among other things.

Fox-McCloud commented on code/modules/clothing/spacesuits/rig/modules/utility.dm:
```diff
@@ -0,0 +1,493 @@
+		list("nutrients",     "nutriment",     0, 80),
+		list("hyronalin",     "hyronalin",     0, 80),
+		list("radium",        "radium",        0, 80)
+		)
```
hardly any of these reagents exist, minus radium and nutriment.

tigercat2000 replied:
Isn't spaceacillin still a thing too?

Fox-McCloud commented on code/modules/clothing/spacesuits/rig/modules/utility.dm:
```diff
@@ -0,0 +1,493 @@
+		list("hyperzine",     "hyperzine",     0, 30),
+		list("oxycodone",     "oxycodone",     0, 30),
+		list("nutrients",     "nutriment",     0, 80),
+		)
```
same here.

Fox-McCloud commented on code/modules/clothing/spacesuits/rig/modules/utility.dm:
```diff
@@ -0,0 +1,493 @@
+		list("nutrients",     "nutriment",     0, 80),
+		list("hyronalin",     "hyronalin",     0, 20),
+		list("radium",        "radium",        0, 20)
+		)
```
same here.

Fox-McCloud commented:
I'd recommend moving (if it's not done so already) all the new rig related files into their own specific folder.

tigercat2000 commented:
It's all under /code/modules/clothing/spacesuits/rig and spacesuits/void

TheDZD commented on code/modules/clothing/spacesuits/rig/modules/utility.dm:
```diff
@@ -0,0 +1,493 @@
+
+	charges = list(
+		list("synaptizine",   "synaptizine",   0, 30),
+		list("morphine",      "morphine",      0, 30),
```
This should probably be hydrocodone, otherwise it's going to be putting people to sleep and making them ignore slows, rather than reducing pain. That said, hydrocodone is incredibly potent and has pretty much no drawbacks, so...

TheDZD commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -256,6 +257,14 @@
 				stat("Tank Pressure", internal.air_contents.return_pressure())
 				stat("Distribution Pressure", internal.distribute_pressure)
+
+		if(back && istype(back, /obj/item/weapon/rig))
```
Checking for `back` should be redundant, considering istype is going to return false if `back` is null.

TheDZD commented on code/modules/mob/living/carbon/human/human.dm:
```diff
@@ -1110,6 +1119,17 @@
+				tinted = 2
+			if((O.offline && O.offline_vision_restriction == 2) || (!O.offline && O.vision_restriction == 2))
+				tinted = 3
+	//im so sorry
```
[Apology accepted?](http://i.imgur.com/DC0H2e7.jpg)

TheDZD commented on code/modules/mob/living/carbon/human/human_movement.dm:
```diff
@@ -58,10 +58,19 @@
-		if((movement_dir || J.stabilization_on) && J.allow_thrust(0.01, src))
+	//Do we have a working jetpack?
+	var/obj/item/weapon/tank/jetpack/thrust
+	if(back)
```
Redundant for the two checks inside of this.

TheDZD commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -62,6 +62,11 @@ var/global/list/brutefireloss_overlays = list("1" = image("icon" = 'icons/mob/sc
+
+	// This is not an ideal place for this but it will do for now.
+	if(wearing_rig && wearing_rig.offline)
+		wearing_rig = null
```
Couldn't this just be done wherever the rig's `offline` var is set to 1 rather than be checked in Life()?

tigercat2000 replied:
Maybe?

TheDZD replied:
Might be a good thing to try then, because Life() code is pretty cluttered as is, and having this check every Life() for every humanoid does add a tiny bit of overhead.

TheDZD commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -1220,6 +1243,32 @@ var/global/list/brutefireloss_overlays = list("1" = image("icon" = 'icons/mob/sc
 						if(ANTAGHUD)
 							process_antag_hud(src)
 
+			if(back && istype(back, /obj/item/weapon/rig)) ///ahhhg so snowflakey
```
[Vision/HUD code makes me cry because of how snowflakey it is.](http://i.imgur.com/DC0H2e7.jpg)

TheDZD commented on code/modules/mob/living/carbon/life.dm:
```diff
@@ -1,8 +1,8 @@
-/mob/living/carbon/proc/get_breath_from_internal(volume_needed)
+/mob/living/carbon/proc/get_breath_from_internal(var/volume_needed=BREATH_VOLUME) //hopefully this will allow overrides to specify a different default volume without breaking any cases where volume is passed in.
```
No reason to add `var/` considering it is implied.

tigercat2000 replied:
copypasted code

TheDZD commented:
Not particularly a fan of some of the snowflaking that hardsuits use, but that is sort of a given with how they function.

Fox-McCloud commented on code/modules/clothing/spacesuits/rig/suits/alien.dm:
```diff
@@ -0,0 +1,32 @@
+	desc = "An authentic Unathi breacher chassis. Huge, bulky and absurdly heavy. It must be like wearing a tank."
+	suit_type = "breacher chassis"
+	icon_state = "breacher_rig"
+	armor = list(melee = 90, bullet = 90, laser = 90, energy = 90, bomb = 90, bio = 100, rad = 80) //Takes TEN TIMES as much damage to stop someone in a breacher. In exchange, it's slow.
```
Not asking you to address this, but seriously Bay? what made you guys think that giving someone 90% damage reduction (and melee immunity with a security jumpsuit) was a remotely sane idea?

tigercat2000 replied:
OH JESUS CHRIST WHAT THE FUCK

Zuhayr replied:
It's an adminspawn item, calm down.

Aurorablade replied:
yeah my legless Unathi needs that one..for reasons....(or at least the IAA tie when i play her as IAA)

Fox-McCloud commented on code/modules/projectiles/guns/energy.dm:
```diff
@@ -23,7 +29,46 @@
+/obj/item/weapon/gun/energy/Destroy()
+	if(self_recharge)
+		processing_objects.Remove(src)
+	..()
```
Return to sender....tell him his parents did call.

=p

Fox-McCloud commented:
Reviewed it, requesting review by @Markolie.

Also uh.....ten bucks says these things don't GC, especially when worn and active, hahaha.

tigercat2000 commented:
Doubt it.

Fox-McCloud commented:
Requesting final review by @Markolie

tigercat2000 commented:
oh god merge conflicts someone kill me now

Fox-McCloud commented:
:cat: :boom: :gun: :dog: 

Alright.

tigercat2000 commented:
thanks

Fox-McCloud commented:
Any time =p

tigercat2000 commented:
i'll see about fixing the conflicts as soon as I push my bodyscanner bodytemperature stuff

tigercat2000 commented:
Oh, easy fix. Just a haunted file.

Regen1 commented:
6 days without any progress, when will this be review?

tigercat2000 commented:
@Regen1 Mark's been busy.

tigercat2000 commented:
oh god merge conflicts

orelbon commented:
well is this merging or no?

TheDZD commented:
@tigercat2000 is working on a Life() refactor, and has told me to hold off on merging this until that goes through, due to conflicts.

tigercat2000 commented:
Merge conflicts fixed, and I actually tested this again, everything (that I can see) works fine.

tigercat2000 commented on code/modules/mob/living/carbon/human/human_damage.dm:
```diff
@@ -130,6 +130,9 @@
 
 /mob/living/carbon/human/Paralyse(amount)
+	// Notify our AI if they can now control the suit.
+	if(wearing_rig && !stat && paralysis < amount) //We are passing out right this second.
```
i only just noticed this
i don't think the AI can actually control the suit

tigercat2000 commented:
All good.

orelbon commented:
margee

TheDZD commented:
Conflicts.

ZomgPonies commented:
Meeeeeeeeeeeerge

TheDZD commented:
:tada:

tigercat2000 commented:
only took 1 1/2 months

orelbon commented:
so this is only admin spawnable right?

tigercat2000 commented:
yes

orelbon commented:
Aw :disappointed:
